iocore/hostdb/P_RefCountCache.h

```diff
@@ -0,0 +1,896 @@
+    this->items--;
+
+    this->metric_inc(refcountcache_current_size_stat, -((int64_t)l.m_value->meta.size));
+    this->metric_inc(refcountcache_current_items_stat, -1);
+
+    // remove from expiry queue
+    if (l.m_value->expiry_entry != NULL) {
+      Debug("refcountcache", "partition %d deleting item from expiry_queue idx=%d\n", this->part_num,
+            l.m_value->expiry_entry->index);
+      this->expiry_queue.erase(l.m_value->expiry_entry);
+      expiryQueueEntry.free(l.m_value->expiry_entry);
+      l.m_value->expiry_entry = NULL; // To avoid the destruction of `l` calling the destructor again-- and causing issues
+    }
+    // Since the Value is actually RefCountObj-- when this gets deleted normally it calls the wrong
+    // `free` method, this forces the delete/decr to happen with the right type
+    Ptr<C> *tmp = (Ptr<C> *)&l.m_value->item;
+    tmp->clear();
+    refCountCacheHashingValueAllocator.free(l.m_value);
+  }
+}
+
+template <class C>
+void
+RefCountCachePartition<C>::clear()
+{
+  this->metric_inc(refcountcache_current_size_stat, -this->size);
+  this->metric_inc(refcountcache_current_items_stat, -this->items);
+  // Clear the in memory hashmap
+  // TODO: delete all items (not sure if clear calls delete on all items)
+  this->item_map.clear();
+  // clear the queue
+  PriorityQueueEntry<RefCountCacheHashEntry *> *tmp_entry;
+  while (!this->expiry_queue.empty()) {
+    tmp_entry = this->expiry_queue.top();
+    this->expiry_queue.pop();
+    expiryQueueEntry.free(tmp_entry);
+  }
+
+  this->items = 0;
+  this->size  = 0;
+}
+
+// Are we full?
+template <class C>
+bool
+RefCountCachePartition<C>::is_full()
+{
+  Debug("refcountcache", "partition %d is full? items %d/%d size %ld/%ld\n\n", this->part_num, this->items, this->max_items,
+        this->size, this->max_size);
+  return (this->max_items > 0 && this->items >= this->max_items) || (this->max_size > 0 && this->size >= this->max_size);
+}
+
+// Attempt to make space for item of `size`
+template <class C>
+bool
+RefCountCachePartition<C>::reserve(unsigned int size)
+{
+  int curr_time = Thread::get_hrtime() / HRTIME_SECOND;
+  while (this->is_full() || (size > 0 && this->size + size > this->max_size)) {
+    PriorityQueueEntry<RefCountCacheHashEntry *> *top_item = expiry_queue.top();
+    // if there is nothing in the expiry queue, then we can't make space
+    if (top_item == NULL) {
+      return false;
+    }
+
+    // If the first item has expired, lets evict it, and then go around again
+    if (top_item->node->meta.expiry_time < curr_time) {
+      this->erase(top_item->node->meta.key);
+      expiry_queue.pop();
+    } else { // if the first item isn't expired-- the rest won't be either (queue is sorted)
+      return false;
+    }
+  }
+  return true;
+}
+
+template <class C>
+const size_t
+RefCountCachePartition<C>::count()
+{
+  return this->items;
+}
+
+template <class C>
+void
+RefCountCachePartition<C>::copy(Vec<RefCountCacheHashEntry *> &items)
```

Comment:
https://issues.apache.org/jira/browse/TS-4600 and https://issues.apache.org/jira/browse/TS-4599